sklearn/model_selection/_split.py

```diff
@@ -0,0 +1,1388 @@
+        the value is automatically set to the complement of the train size.
+
+    train_size : float, int, or None (default is None)
+        If float, should be between 0.0 and 1.0 and represent the
+        proportion of the dataset to include in the train split. If
+        int, represents the absolute number of train samples. If None,
+        the value is automatically set to the complement of the test size.
+
+    random_state : int or RandomState
+        Pseudo-random number generator state used for random sampling.
+
+    Examples
+    --------
+    >>> from sklearn.model_selection import StratifiedShuffleSplit
+    >>> X = np.array([[1, 2], [3, 4], [1, 2], [3, 4]])
+    >>> y = np.array([0, 0, 1, 1])
+    >>> sss = StratifiedShuffleSplit(n_iter=3, test_size=0.5, random_state=0)
+    >>> sss.get_n_splits(X, y)
+    3
+    >>> print(sss)       # doctest: +ELLIPSIS
+    StratifiedShuffleSplit(n_iter=3, random_state=0, ...)
+    >>> for train_index, test_index in sss.split(X, y):
+    ...    print("TRAIN:", train_index, "TEST:", test_index)
+    ...    X_train, X_test = X[train_index], X[test_index]
+    ...    y_train, y_test = y[train_index], y[test_index]
+    TRAIN: [1 2] TEST: [3 0]
+    TRAIN: [0 2] TEST: [1 3]
+    TRAIN: [0 2] TEST: [3 1]
+    """
+
+    def __init__(self, n_iter=10, test_size=0.1, train_size=None,
+                 random_state=None):
+        super(StratifiedShuffleSplit, self).__init__(
+            n_iter, test_size, train_size, random_state)
+
+    def _iter_indices(self, X, y, labels=None):
+        n = _num_samples(X)
+        n_train, n_test = _validate_shuffle_split(n, self.test_size,
+                                                  self.train_size)
+        classes, y_indices = np.unique(y, return_inverse=True)
+        n_cls = classes.shape[0]
+
+        if np.min(bincount(y_indices)) < 2:
+            raise ValueError("The least populated class in y has only 1"
+                             " member, which is too few. The minimum"
+                             " number of labels for any class cannot"
+                             " be less than 2.")
+
+        if n_train < n_cls:
+            raise ValueError('The train_size = %d should be greater or '
+                             'equal to the number of classes = %d' %
+                             (n_train, n_cls))
+        if n_test < n_cls:
+            raise ValueError('The test_size = %d should be greater or '
+                             'equal to the number of classes = %d' %
+                             (n_test, n_cls))
+
+        rng = check_random_state(self.random_state)
+        cls_count = bincount(y_indices)
+        p_i = cls_count / float(n)
+        n_i = np.round(n_train * p_i).astype(int)
+        t_i = np.minimum(cls_count - n_i,
+                         np.round(n_test * p_i).astype(int))
+
+        for n in range(self.n_iter):
+            train = []
+            test = []
+
+            for i, cls in enumerate(classes):
```

Comment:
but I am unable to think of a better name other than `class` maybe `class_number` / `class_n` / `class_` - @arjoly @amueller